internal/auth/service_list_auth_methods_refresh.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package auth
+
+import (
+	"context"
+	"slices"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/refreshtoken"
+)
+
+func (s *AuthMethodService) ListAuthMethodsRefresh(
+	ctx context.Context,
+	tok *refreshtoken.Token,
+	scopeIds []string,
+	withUnauthenticatedUser bool,
+	pageSize int,
+	grantsHash []byte,
+	filterItemFn func(AuthMethod) (bool, error),
+) (*ListAuthMethodsResponse, error) {
+	const op = "auth.(*AuthMethodService).ListAuthMethodsRefresh"
+
+	// Request and combine deleted ids from the DB for ldap, oidc, and passwords.
+	var deletedIds []string
+	var transactionTimestamp time.Time
+	if _, err := s.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{}, func(r db.Reader, w db.Writer) error {
+		deletedLdapIds, err := s.ldapRepo.ListDeletedAuthMethodIds(ctx, tok.UpdatedTime, WithReaderWriter(ctx, r, w))
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		deletedOidcIds, err := s.oidcRepo.ListDeletedAuthMethodIds(ctx, tok.UpdatedTime, WithReaderWriter(ctx, r, w))
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		deletedPwIds, err := s.passwordRepo.ListDeletedAuthMethodIds(ctx, tok.UpdatedTime, WithReaderWriter(ctx, r, w))
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		transactionTimestamp, err = r.Now(ctx)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		deletedIds = append(deletedLdapIds, append(deletedOidcIds, deletedPwIds...)...)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	limit := pageSize + 1
+	opts := []Option{
+		WithLimit(ctx, limit),
+		WithStartPageAfterItem(ctx, tok.LastItemId, tok.LastItemUpdatedTime),
+	}
+	if withUnauthenticatedUser {
+		opts = append(opts, WithUnauthenticatedUser(ctx, withUnauthenticatedUser))
+	}
+
+	authMethods := make([]AuthMethod, 0, limit)
+dbLoop:
+	for {
+		// Request pages from the DB for ldap, oidc, and passwords.
+		// Combine, sort, and fill the pages until we fill the final items
+		ldapPage, err := s.ldapRepo.ListAuthMethods(ctx, scopeIds, opts...)
+		if err != nil {
+			return nil, errors.Wrap(ctx, err, op)
+		}
+		oidcPage, err := s.oidcRepo.ListAuthMethods(ctx, scopeIds, opts...)
+		if err != nil {
+			return nil, errors.Wrap(ctx, err, op)
+		}
+		pwPage, err := s.passwordRepo.ListAuthMethods(ctx, scopeIds, opts...)
+		if err != nil {
+			return nil, errors.Wrap(ctx, err, op)
+		}
+		finalPage := append(ldapPage, append(oidcPage, pwPage...)...)
+		slices.SortFunc(finalPage, func(i, j AuthMethod) int {
+			return i.GetUpdateTime().AsTime().Compare(j.GetUpdateTime().AsTime())
+		})
+		if len(finalPage) > limit {
+			finalPage = finalPage[:limit]
+		}
```

Comment:
This is a little annoying as we're requesting 3x the size of each page, but I'm not sure there's another way to do it. This logic guarantees that we'll get the correct order of auth methods, even if all of them are in one particular auth method.